src/cylinder.jl

```diff
@@ -25,12 +25,21 @@ function Base.in(x::SVector{N}, s::Cylinder{N}) where {N}
     return sum(abs2,d - p*s.a) ≤ s.r^2
 end
 
-function normal(x::SVector{N}, s::Cylinder{N}) where {N}
+function surfpt_nearby(x::SVector{N}, s::Cylinder{N}) where {N}
     d = x - s.c
-    p = dot(d, s.a)
-    p > s.h2 && return s.a
-    p < -s.h2 && return -s.a
-    return normalize(d - p*s.a)
+    p = d ⋅ s.a
+    q = d - p*s.a
+    lp = abs(p)
+    lq = norm(q)
+    if abs(lp-s.h2) < abs(lq-s.r)
+        nout = sign(p)*s.a
+        l∆x = s.h2 - lp
+    else
+        l∆x = s.r - lq
+        nout = q / lq
```

Comment:
The reason I didn't worry too much about `lq` being zero was that I thought we would only care about `x` close to the surface.  If `lq == 0`, `x` is very apart from the side wall of the cylinder, so for such `x` to be close to the surface it must be close to one of the bases (in other words `abs(lp-s.h2) < abs(lq-s.r)` is true and the `elseif` block containing the division by `lq` is not executed).

I wouldn't mind handling the `lq == 0` case for safety, but I'm just curious if there are any practical cases